snuba/query/snql/parser.py

```diff
@@ -1059,6 +1060,59 @@ def validate_entities_with_query(
                     )
 
 
+def _select_entity_for_dataset(
+    dataset: Dataset,
+) -> Callable[[Union[CompositeQuery[QueryEntity], LogicalQuery]], None]:
+    def selector(query: Union[CompositeQuery[QueryEntity], LogicalQuery]) -> None:
+        # If you are doing a JOIN, then you have to specify the entity
+        if isinstance(query, CompositeQuery):
+            return
+
+        if get_dataset_name(dataset) == "discover":
+            query_entity = query.get_from_clause()
+            # The legacy -> snql parser will mark queries with no entity specified as the "discover" entity
+            # so only do this selection in that case. If someone wants the "discover" entity specifically
+            # then their query will have to only use fields from that entity.
+            if query_entity.key == EntityKey.DISCOVER:
+                selected_entity_key = dataset.select_entity(query)
+                selected_entity = get_entity(selected_entity_key)
+                query_entity = QueryEntity(
+                    selected_entity_key, selected_entity.get_data_model()
+                )
+                query.set_from_clause(query_entity)
+
+                # XXX: This exists only to ensure that the generated SQL matches legacy queries.
+                def replace_time_condition_aliases(exp: Expression) -> Expression:
+                    if (
+                        isinstance(exp, FunctionCall)
+                        and len(exp.parameters) == 2
+                        and isinstance(exp.parameters[0], Column)
+                        and exp.parameters[0].alias == "_snuba_timestamp"
+                    ):
+                        return FunctionCall(
+                            exp.alias,
+                            exp.function_name,
+                            (
+                                Column(
+                                    f"_snuba_{selected_entity.required_time_column}",
+                                    exp.parameters[0].table_name,
+                                    exp.parameters[0].column_name,
+                                ),
+                                exp.parameters[1],
+                            ),
+                        )
+
+                    return exp
```

Comment:
I am not sure I follow the rationale of this transformation. What is the discrepancy it is addressing ?

Reply:
This is for the specific case of when a discover query gets routed to transactions. For legacy queries, the time series extension runs after entity selection and knows the correct time field for the entity and generates an column with an alias of the form “finish_ts AS _snuba_finish_ts” when it adds the time column conditions.

The SDK adds those conditions before the entity selection, so the column changes from timestamp to finish_ts. Changing the column works fine, however the alias doesn’t change so the condition column looks like “finish_ts AS _snuba_timestamp”. Since the aliases don’t match, the SQL comparison fails.

Reply:
I think I’m going to add a special “legacy” parameter to the SnQL json Schema, so that I can flag logic like this for the specific case of legacy conversion. That way it will be easier to track down once the migration is complete.

Reply:
I hope this will go away.

Reply:
It will ... eventually.